core/corehttp/hostname.go

```diff
@@ -266,6 +290,37 @@ func isPeerIDNamespace(ns string) bool {
 	}
 }
 
+// Converts an identifier to DNS-safe representation
+func toDNSPrefix(id string) (prefix string) {
+	s := strings.Replace(id, ".", "", -1) // remove separators if present
+
+	// Return if things fit after dot removal
+	if len(s) <= dnsLabelMaxLength {
+		return s
+	}
+
+	parts := make(
+		[]string,
+		// same as ceil( len(s) / dnsLabelMaxLength )
+		(len(s)+dnsLabelMaxLength-1)/dnsLabelMaxLength,
+	)
+
+	firstPartLen := len(s) % dnsLabelMaxLength
+
+	// if it divides by 63 perfectly - full part
+	if firstPartLen == 0 {
+		firstPartLen = dnsLabelMaxLength
+	}
+
+	// Iterate from right to left to maximize length of right-most labels
+	for i := len(parts) - 1; i > 0; i-- {
+		parts[i] = s[(i-1)*dnsLabelMaxLength+firstPartLen : i*dnsLabelMaxLength+firstPartLen]
+	}
+	parts[0] = s[:firstPartLen]
```

Comment:
Wouldn't you want to maximize the left-most labels? That is, make the right-most labels as small as possible?

We're only over by two base-32 numerals so we can _probably_ get `32**2 = 1024` wildcard certificates. It looks like we can register 100 names per cert, and 50 certs which gives us 5000 certs.

Reply:
We wanted to maximize the first part after [public suffix](https://publicsuffix.org), but you have a point, it may be possible to get TLS certs if we hardcode the first split to be after two characters from the right. 

My main concern: while technically possible, I am not sure if it is practical / better than using Base36 for ED25519.

Assuming we have all the certs in place, this fix would work only for ED25519 represented in base32, everything else (eg. sha2-512)  will still be over the wildcard length limit:

- :+1: ED25519 libp2p-key is only 2 characters over the limit, so it fits in two labels: https://bafzaajaiaejca4syrpdu6gdx4wsdnokxkprgzxf4wrstuc34gxw5k5jrag2so5.gk.ipns.dweb.link
  - `bafzaajaiaejca4syrpdu6gdx4wsdnokxkprgzxf4wrstuc34gxw5k5jrag2so5` . `gk`
- :anger:  CID created with `--hash sha2-512` will be too long, needs to be split into more than two chunks anyway: https://bafkrgqe3ohjcjplc6n4f3fwunlj6upltggn7xqujbsv.nvyw764srszz4u4rshq6ztos4chl4plgg4ffyyxnayrtdi5oc4xb2332g645433a.eg.ipfs.dweb.link:
  - `bafkrgqe3ohjcjplc6n4f3fwunlj6upltggn7xqujbsv`  . `nvyw764srszz4u4rshq6ztos4chl4plgg4ffyyxnayrtdi5oc4xb2332g645433a` . `eg`


 
Let's Encrypt limits are ([src](https://community.letsencrypt.org/t/wildcard-certificate-limit/86536/2)):
> - You can only fit 100 domains onto one certificate
> - Each Registered Domain may only appear on 50 certificates per week.

Interesting part is that according to the [rate limits](https://letsencrypt.org/docs/rate-limits/) they

> use the Public Suffix List to calculate the registered domain. 

`ipfs.dweb.link` and `ipns.dweb.link` are both on Public Suffix List, which means they are "excluded" from limit calculation, and the first CID chunk becomes "Registered Domain", effectively removing the weekly limit. 

I'd appreciate input from @mburns and @MichaelMure on how feasible it is for us, or operator like Infura to get around 100 certs, each for 100 wildcard names to support TLS on `*.aa.dweb.link`, `*.ab.dweb.link` ...  etc, **and if it's better than using Base36 and not splitting IPNS ED25519 names.**

Thoughts?

Reply:
I found out about this public suffix "loophole" earlier today as well. It could make things easier, but it's also somewhat abusive and risky (do they even update their list?). There is also the possibility to apply for a raised rate limit but I don't know of feasible that is. Not even talking about the vendor locking with LetsEncrypt.

One problem coming from this solution is also that it pretty much means doing TLS termination manually. AWS for example limit to 25 certificates attached to a load balancer. It compound quickly in complexity.

Sadly we might need to go that way anyway because we need to have a user ID as a subdomain (we need to link requests to a user somehow, basic auth break some usecases, we can't touch the path part of the URL).

In any case, base36 is way way easier for us.

Reply:
I agree with using base36. And you're right, _this_ feature is just to support sha512 etc. so it doesn't really matter.

> We want to maximize the first part after public suffix, giving the most bits to parent labels, following the spirit of DNS hierarchy.

I'm not sure I follow. Why would we want to maximize that part? Are we worried about cookies/origins?

Reply:
Managing that many certs is doable, but feels a bit icky, technically speaking.

all else being equal, sounds like base36 is preferable.

Reply:
> Why would we want to maximize that part? Are we worried about cookies/origins?

AFAIK in practice, it does not matter. 
But if we can pick, I'd maximize "parent" labels on the right, as a _precaution_.  

My rationale: the surface for bugs always exists.
To illustrate: a mild version of "Origin sharing" is a thing. Two sibling subdomains can _mutually_ agree to use parent Origin for cookies (_both_ setting `document.domain = aa.ipfs.example`, assuming `example.com` is not on Public Suffix List). 

Let's say the future brings a bug/vulnerability in Origin-separation code in one of browser vendors. Maximizing parent label makes it harder/infeasible to pull this class of attacks off, as generating parent label if way more difficult if it needs to match 63 instead of 2 chars.

Reply:
TLDR: seconding what @lidel said

In fact this very thread contains something that walks and talks like a precursor to a phishing attack: the ability to blanket register a star-TLS cert to all possible N-character prefixes when N is short enough at the top.

Before this thread started this was also one of the 1st questions I asked @lidel when I was doing review, because I was dumb and misread the code: "@lidel why do you let the rightmost side be short and thus brute-forceable?"

I do not have a good feeling how to "productize" this into an outright vulnerability, but leaving a "mid-to-top-level" part of DNS trivially forgeable doesn't... smell right at all.